### ISSUE TITLE
Retrying connection to AMQP broker

### ISSUE THREAD

**Issue by s-maj** [cloudamqp/amqproxy#14]
Looks like I can authenticate now but connection is lost after a while, most likely during send or receive. 

```
Unhandled exception in spawn: AMQ::Protocol::Frame::Body-end was 123, expected 206 (AMQ::Protocol::Error::InvalidFrameEnd)
  from /usr/share/crystal/src/gc/boehm.cr:71:5 in '->'
  from /usr/share/crystal/src/fiber.cr:255:3 in '???'
  from ???
```
```
Unhandled exception in spawn: AMQ::Protocol::Frame::Body-end was 10, expected 206 (AMQ::Protocol::Error::InvalidFrameEnd)
  from /usr/share/crystal/src/gc/boehm.cr:71:5 in '->'
  from /usr/share/crystal/src/fiber.cr:255:3 in '???'
  from ???
```
```
Unhandled exception in spawn: AMQ::Protocol::Frame::Body-end was 116, expected 206 (AMQ::Protocol::Error::InvalidFrameEnd)
  from /usr/share/crystal/src/gc/boehm.cr:71:5 in '->'
  from /usr/share/crystal/src/fiber.cr:255:3 in '???'
  from ???
```
```
Unhandled exception in spawn: AMQ::Protocol::Frame::Body-end was 99, expected 206 (AMQ::Protocol::Error::InvalidFrameEnd)
  from /usr/share/crystal/src/gc/boehm.cr:71:5 in '->'
  from /usr/share/crystal/src/fiber.cr:255:3 in '???'
  from ???
```
```
Unhandled exception in spawn: AMQ::Protocol::Frame::Body-end was 115, expected 206 (AMQ::Protocol::Error::InvalidFrameEnd)
  from /usr/share/crystal/src/gc/boehm.cr:71:5 in '->'
  from /usr/share/crystal/src/fiber.cr:255:3 in '???'
  from ???
```
```
Unhandled exception in spawn: AMQ::Protocol::Frame::Body-end was 105, expected 206 (AMQ::Protocol::Error::InvalidFrameEnd)
  from /usr/share/crystal/src/gc/boehm.cr:71:5 in '->'
  from /usr/share/crystal/src/fiber.cr:255:3 in '???'
  from ???
```
```
Unhandled exception in spawn: Could not write the full body (AMQ::Protocol::Error::FrameEncode)
  from lib/amq-protocol/src/amq/protocol/properties.cr:136:23 in '???'
  from src/amqproxy/server.cr:112:13 in '???'
  from src/amqproxy/pool.cr:17:7 in '->'
  from /usr/share/crystal/src/fiber.cr:255:3 in '???'
  from ???
```

**Comment by carlhoerberg** [comment]
With which client was this? After continuous publishing? Doing something else on the connection? The more information the easier it will be to replicate, and then fix.

**Comment by s-maj** [comment]
Right, this is  mostly Celery(4.2.1)/Kombu(4.2.1)/PyAMQP(2.3.2) and one Pika app (latest version) It looks like apps can connect but cannot send or receive any messages. 

For the Pika:
`Connection close detected; result=BlockingConnection__OnClosedArgs(connection=<SelectConnection CLOSED socket=None params=<URLParameters host=10.240.9.151 port=9999 virtual_host=platform-stress ssl=False>>, reason_code=505, reason_text='UNEXPECTED_FRAME - expected content header for class 60, got non content header frame instead')`

For Celery/py-amqp I think it's about heartbeats that being send by default by producer (it can't be disabled atm):
`Unrecoverable error: ConnectionForced(None, u'Too many heartbeats missed', None, u'')`

but also are some ambiguous messages like:
`CRITICAL/MainProcess] Unrecoverable error: ConnectionError(302, u'UPSTREAM_ERROR', (0, 0), u'')`

**Comment by carlhoerberg** [comment]
I know what the issue is, I'm working on a fix, i think it will be ready tmrw. https://github.com/cloudamqp/amqproxy/releases/tag/v0.2.3 does not have the same issue as the current 0.3.x line.

**Comment by carlhoerberg** [comment]
Fixed in v0.3.3: https://github.com/cloudamqp/amqproxy/releases/tag/v0.3.3
